drivers/can/can_renesas_ra.c

```diff
@@ -0,0 +1,1171 @@
+		return -EBUSY;
+	}
+
+	can_bit_timing_cfg_t data_timing = ZEPHYR_TO_FSP_TIMING(timing_data);
+
+	if (!r_canfd_bit_timing_parameter_check(&data_timing, true)) {
+		return -ENOTSUP;
+	}
+	data->custom.data_timing = data_timing;
+
+	return 0;
+}
+#endif /* CONFIG_CAN_FD_MODE */
+
+/**************************************************************************************************
+ * Internal function definition
+ *************************************************************************************************/
+void can_renesas_ra_fsp_cb(can_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+
+	switch (p_args->event) {
+	case CAN_EVENT_RX_COMPLETE: {
+		struct can_frame z_frame = {0};
+
+		FSP_TO_ZEPHYR_FRAME(&p_args->frame, &z_frame);
+		can_renesas_ra_call_rx_cb(dev, &z_frame);
+		break;
+	}
+
+	case CAN_EVENT_TX_COMPLETE: {
+		can_renesas_ra_call_tx_cb(dev);
+		break;
+	}
+
+	case CAN_EVENT_ERR_CHANNEL: {
+		if (p_args->error & R_CANFD_CFDC_ERFL_BEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_ACTIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EWF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_WARNING);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EPF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_PASSIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_BOEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_BUS_OFF);
+		}
+		break;
+	}
+
+	default:
+		break;
+	}
+}
+
+static inline void can_renesas_ra_call_tx_cb(const struct device *dev)
+{
+	struct can_renesas_ra_data *data = dev->data;
+
+	data->custom.tx_cb(dev, 0, data->custom.tx_usr_data);
+	k_work_cancel_delayable(&data->custom.tx_timeout_work);
+}
+
+static inline void can_renesas_ra_call_rx_cb(const struct device *dev, struct can_frame *rx_frame)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	const struct can_renesas_ra_cfg *cfg = dev->config;
+	struct can_renesas_ra_filter *rx_filter = data->custom.rx_filter;
+
+	for (uint32_t i = 0; i < cfg->custom.rx_filter_num; i++) {
+		if (!rx_filter->rx_cb) {
+			rx_filter++;
+			continue; /* this filter has not set yet */
+		}
+
+		if (!can_frame_matches_filter(rx_frame, &rx_filter->filter)) {
+			rx_filter++;
+			continue; /* filter did not match */
+		}
+
+		rx_filter->rx_cb(dev, rx_frame, rx_filter->rx_usr_data);
+		break;
+	}
+}
+
+static inline void can_renesas_ra_call_state_change_cb(const struct device *dev,
+						       enum can_state state)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	can_info_t can_info = {};
+	fsp_err_t fsp_err;
+
+	fsp_err = can_api->infoGet(data->custom.fsp_can.p_ctrl, &can_info);
+	if (fsp_err != FSP_SUCCESS) {
+		LOG_DBG("drivers: can: can get err info failed");
+		return;
+	}
+	struct can_bus_err_cnt err_cnt = {
+		.rx_err_cnt = can_info.error_count_receive,
+		.tx_err_cnt = can_info.error_count_transmit,
+	};
+
+	if (data->common.state_change_cb == NULL) {
```

Comment:
If the state change callback is not set, why not disable this interrupt all together and save some CPU cycles?

Reply:
I fixed it. State change interrupts will be disabled until the callback is set